DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,927 @@
+of fields which makes it impossible to type check and cannot be synchronized without
+additional costs.
+
+As an alternative, this DIP proposes the implementation of a copy constructor. The benefits
+of this approach are the following:
+
+* it is a known concept. C++ has it [3];
+* it can be typechecked as a normal constructor (since no blitting is done, the data is
+initialized as if it were in a normal constructor); this means that `const`/`immutable`/`shared`
+copy constructors will be type checked exactly as their analogous constructors
+* offers encapsulation
+
+The downside of this solution is that the user must do all the field copying by hand
+and every time a field is added to a struct, the copy constructor must be modified.
+However, this can be easily bypassed by D's introspection mechanisms. For example,
+this simple code may be used as a language idiom or library function:
+
+```D
+foreach (i, ref field; src.tupleof)
+    this.tupleof[i] = field;
+```
+
+As shown above, the single benefit of the postblit can be easily substituted with a
+few lines of code. On the other hand, to fix the limitations of the postblit it is
+required that more complexity is added for little to no benefit. In these circumstances,
+for the sake of uniformity and consistency, replacing the postblit constructor with a
+copy constructor is the reasonable thing to do.
+
+## Description
+
+This section discusses all the technical aspects regarding the semantics
+of the copy constructor.
+
+### Syntax
+
+The following link exhibits the proposed syntax and the necessary grammar changes in order to support
+the copy construct:
+
+https://github.com/dlang/dlang.org/pull/2414
+
+A declaration is a copy constructor declaration if it is a constructor declaration annotated with
+the `@implicit` attribute and takes only one parameter by reference that is of the same type as
+`typeof(this)`. `@implicit` is a compiler recognised attribute like `@safe` or `nogc` and it can be
+syntactally used the same as the others. `@implicit` can be legally used solely to mark the declaration
+of a copy constructor; all other uses of this attribute will result in a compiler error. The proposed syntax
+benefits from the advantage of declaring the copy constructor in an expressive manner without adding
+additional complexity to the existing syntax.
+
+The copy constructor needs to be annotated with `@implicit` in order to distinguish a copy
+constructor from a normal constructor and avoid silent modification of code behavior. Consider
+this example:
+
+```d
+import std.stdio;
+struct A {
+    this(ref immutable A obj) { writeln("x"); }
+}
+void main()
+{
+    immutable A ia;
+    A a = A(ia);
+    A b = ia;
+}
+```
+
+With the current state of the language, `x` is printed once. with the addition of the copy constructor without
+`@implicit`, `x` would be printed twice, thus modifying the semantics of existing code.
```

Comment:
Is `@implicit` really necessary?

Put other ways:
Is the breakage due to the reservation of `@implicit` worse than modifying the semantics of existing code (especially if transitioned with a `-dip10xx` flag)?
If someone is declaring a constructor of signature `this( ref [qualified] typeof(this) a) [qualified]`, is there any case that they _don't_ want it to be a copy constructor given that the copy constructor is to replace the postblit (again with transition)?

Reply:
`@implicit` doesn’t need to be reserved, it can be a compiler recogized UDA.

Reply:
I'm saying it doesn't need to exist, i.e. `@implicit` can itself, be implicit under a compiler switch.

Reply:
Honestly, I would expect using `@implicit` (or any other attribute) for copy constructors would ultimately cause more bugs than it would ever prevent. Very few functions that would qualify as copy constructors if they were marked with `@implicit` will have been declared in D code at this point, because it really doesn't make much sense to (that's what postblit constructors are for), and if they were declared, does anyone honestly expect them to be intended to work as anything other than copy constructors (albeit ones that must currently be called explicitly)? While not requiring `@implicit` _might_ break existing code, it's not particularly likely, and long term, it would be far better to have a transitional compiler switch than to require `@implicit` just for fear of breaking code that is likely to be extremely rare (if it exists at all) and which likely would work perfectly well with the changes anyway.

On the other hand, if we require `@implicit`, then it's bound to become a common bug to forget `@implicit`, resulting in code that you _think_ is using a custom copy constructor but in fact is using the default, because you forgot `@implicit`. So, we're talking about trying to avoid breaking functions that are likely to be extremely rare by using a solution that's pretty much guaranteed to cause plenty of bugs in the future. I agree that we should be trying to avoid breaking exist code, but that tradeoff is just plain horrible. If we don't want to just immediately, silently break extremely rare code, then providing a compiler switch initially like we have with other DIPs would make a lot more sense. It might even make more sense anyway if we want a chance to bang on the implementation of copy constructors first to make sure that they're working properly before forcing them on everyone.

Reply:
>and if they were declared, does anyone honestly expect them to be intended to work as anything other than copy constructors (albeit ones that must currently be called explicitly)

I think that this is a major point in the need of @implicit. When the user defined the constructor (even though it was intended to work as a copy constructor) he/she didn't have any expectation that it would get called implicitly. The fact that the function was designed to be called explicitly is the reason why @implicit is needed.More than that, there might be users out there that want a copy constructor which is called explicitly and one that is called implicitly (I can't think of any use case, but regardless of that, there might be folks that want it); by not requiring `@implicit` there is no way they can achieve it.

>On the other hand, if we require @implicit, then it's bound to become a common bug to forget @implicit

This is mere speculation.

> So, we're talking about trying to avoid breaking functions that are likely to be extremely rare by using a solution that's pretty much guaranteed to cause plenty of bugs in the future.

When the people will read the changelog and see that the copy constructor feature is now available, I'm guessing that they are going to read the specification first and so copy construction is going to be associated with `@implicit`.

Other pros for @implicit:
 
 -> it is a lot easier to spot copy constructors and differentiate them from normal constructors when reading code when you see `@implicit`
-> it will be easier to extend the feature to copy construct from any type to any type.
-> it brings more clarity (you know that that specific constructor is used implicitly as opposed to an unmarked copy constructor which it's a bit confusing if  it is used implicitly or explicitly)

Reply:
> didn't have any expectation that it would get called implicitly. 

That is the entire point of the transitional process.

> there might be users out there that want a copy constructor which is called explicitly and one that is called implicitly

This is also speculation.

>>On the other hand, if we require @implicit, then it's bound to become a common bug to forget @implicit
>This is mere speculation.

In contrast to all the other attributes they can be ignored and the code is unaffected or the compiler emits an error e.g. I forget `nothrow`: either is doesn't matter (because the code that calls it is not `nothrow), or it is and the compiler complains. With this I forget `@implicit`, and the code is wrong and still compiles.

>  it is a lot easier to spot copy constructors and differentiate them from normal constructors when reading code when you see @implicit

Is it really that much worse than for postblits? Is that even a problem?

> it will be easier to extend the feature to copy construct from any type to any type.

Is this a thing we are even considering?

> it brings more clarity

It does. Is it needed?

Reply:
>Is this a thing we are even considering?

Yes.

Reply:
>> Is this a thing we are even considering?
> Yes.

💛 !!!

On `@implicit`: I really like the explicit part of being implicit. 

But this analysis:

> On the other hand, if we require @implicit, then it's bound to become a common bug to forget @implicit, resulting in code that you think is using a custom copy constructor but in fact is using the default, because you forgot @implicit.

I think will be quite true once you add another part of this DIP:

> Whenever an initialization by copying is encountered, a call to a copy constructor is tried
and if a matching copy constructor is found it is used, otherwise if no copy constructor is
suitable, standard copying is employed

Experience says that people forget extra configurations, so this will happen if the compiler does the standard copying by default. People will think they've defined a copy ctor.

Is it possible to only perform standard copying *iff* an implicit copy constructor is actually declared. This way you'll get a compiler error : `cannot implicitly copy type, please add an @implicity copy contructor` or something. Is that possible?

Or maybe the postblit makes that a bit difficult :/ 

I guess a transition flag may help here ;)

Reply:
>It is not speculation,  

It is, you missed 

> there _might_ be users out there that want

(emphasis mine), you also provided no use case for it.

>> Is this a thing we are even considering?
> Yes.

Oh dear (and you thought `@implicit` was contentious...) . Regardless, that is not part of _this_ DIP; this DIP is replacing something that is implicit (postblit), with something else that is implicit (copy construction).

Reply:
I don't see how anyone could argue that it's not going to be a common thing to forget to use `@implicit`. It doesn't actually add anything. All it does is turn on the behavior that everyone is going to expect. Sure, over time, folks will get used to it and be better at remembering it, but I can guarantee that many folks will forget it just like they forget other attributes. However, in this case, instead of getting a compiler warning because you did something like call an `@system` function from an `@safe` function, because you forgot to mark the function with `@safe`, you'll end up with the default copying mechanism silently being used instead of your custom copy constructor, which may or may not quickly be caught. So, if we ignore the issue of possibly breaking existing code, _all_ `@implicit` does is provide us an easy way to have extra bugs.

There really is no reason to be declaring a copy constructor if you don't intend for it to be implicit. So, _every_ copy constructor is going to need the attribute, making it so that it's simply a source of bugs (when it's forgotten) rather than a benefit. The only reason to have `@implicit` is to avoid breaking code which is almost certainly extremely rare. We've made breaking changes in the past like this by using compiler switches to enable the transition. Why don't we simply do that here as well? I see _no_ benefit to having `@implicit` over a transitional compiler switch and several negatives. Adding `@implicit` for fear of breaking rare code just means that we're trying to avoid short term breakage at the cost of making the language worse in the long term - and we _already_ have a way around this that we've used with other DIPS. So, let's _please_ use a transitional compiler switch rather than introducing `@implicit` for copy constructors. `@implicit` is just going to make the language worse.